docs/design/referral-manager.md

```diff
@@ -0,0 +1,206 @@
+
+## Background
+
+Along with cross-satellite referrals, we are also working on having a sign-up page that will allow users to select which satellite
+they would like to register an account with. Below design is built based on the new sign-up flow.
+
+## Design
+
+####Definition
+*Referral Link* - _A static link that contains a unique referral code_
+
+*Referral Code* - _An uuid generated by referral manager program for each user to be able to participate in our referral program_
+
+#### User Flow
+1. Sign up
+
+    After new user Alice chose to register on satellite B and clicked on the create account button:
+    - Satellite B sends a request with its satellite id to referral manager for a referral code for Alice
+    - Referral manager receives the request from satellite B, generates an unique referral code 11111111, and then store that code, 11111111, along with the satellite B's address into referral manager's database 
+    - After successfully storing the referral code and satellite B's address, referral manager sends a response with the generated referral code, 11111111, back to satellite B 
+    - Satellite B receives the referral code, 11111111, from referral manager and then store it along with Alice's information into user table in satellitedb
+        - Signed up WITHOUT referral link:
+            - Satellite B will store Alice's credit into user_credit table with a null value in referred_by column
+        - Signed up WITH referral link:
+            - If user Bonnie from satellite C referred Alice with a referral link, https://C.tardigrade.io/signup/22222222
+            - satellite B will store the referral code, 22222222, in user_credits table under referred_by column along with Alice's credits information.
+
+2. Award referrer credits
+    
+    After user Alice has used up their free credits:
+    
+    - Satellite B will send a request with the referral code, 22222222, to referral manager
+    - Referral manager receives the request from satellite B and use the referral code, 22222222, to look up the satellite address that's associated with the referral code, 22222222, in its database
+        - If no referral code can be found in referral manager's database that's equal to 22222222
+            - referral manager sends back an INVALID response to satellite B
+        
+        - If a satellite address, C, is found through the referral code 22222222
+            - referral manager sends the referral code 22222222 to satellite C
+            - referral manager sends back a OK response to satellite B
+            - satellite C receives the referral code 22222222
+            - satellite C looks up the user Bonnie that's associated with the referral code 22222222
+            - satellite C injects a new entry into user_credits table for Bonnie for their earned referrer credits
+            
+## Rationale
+1. Will this overload satellites?
+2. Should we just have satellites communicate with each other directly?
```

Comment:
No, that creates more problems than a single "service", like our bootstrap service currently.

Reply:
@egonelbre I have listed having the satellites communicating with each other directly as an question but seems like most people are against it

Reply:
Yeah, if we intend to add more responsibilities to the referral manager then, yes, it makes sense.